python/lsst/cp/pipe/ptc/cpExtractPtcTask.py

```diff
@@ -257,9 +257,45 @@ def runQuantum(self, butlerQC, inputRefs, outputRefs):
             inputs['inputExp'] = arrangeFlatsByExpId(inputs['inputExp'], inputs['inputDims'])
 
         outputs = self.run(**inputs)
+        outputs = self._guaranteeOutputs(inputs['inputDims'], outputs, outputRefs)
         butlerQC.put(outputs, outputRefs)
 
+    def _guaranteeOutputs(self, inputDims, outputs, outputRefs):
```

Comment:
Two questions: 
1) Why do we need this? I thought that `cpPtcExtract` was already introducing dummy exposures in order to match the inputs: https://github.com/lsst/cp_pipe/blob/main/python/lsst/cp/pipe/ptc/cpExtractPtcTask.py#L311 
or are you building a pipeline that will only use the classes in `cpPtcSolve` without running `cpPtcExtract` first? 
2) It's not that clear to me how the function is implementing the matching.

Reply:
The issue arises when running in the `cp_testing` environment.  There is a check during the ISR stage to restrict the inputs to flat exposures only, returning `pipeBase.NoWorkFound` for the other exposures.  This is fine in other stages, such as flat combine, as whatever exposures are passed as input are used to construct a single output.  However, `cpPtcExtract` is expecting to write an output for all inputs to the pipeline, so it ends up in a situation where (for my test case) `len(inputRefs.inputExp) = 124` but `len(outputRefs.outputCovariances) = 229`.  The existing code can supply a dummy dataset for each input, but that leaves 105 expected outputs that are not generated.  This new function iterates over the full set of `outputRefs.outputCovariances`, constructing the missing outputs and inserting them into the expected output location.
As for the matching:  the function iterates over all of the outputs, and if that output's `exposureId` is found in the `inputDims` list, then it needs to extract that entry from the existing outputs (which are currently indexed by the ordering of `exposureId` in `inputDims`) and insert it into the location expected in the list of `outputRefs`.  If the expected output is not in the existing output list, then the input exposure was ignored during ISR, and so a dummy dataset needs to be inserted into that location.

Reply:
How do you end up with `len(outputRefs.outputCovariances) = 229` after running `cpPtcExtract`, if the input number was only `124`?